### DISCUSSION TITLE
Autoscroll to bottom not working when typing command and output

### DISCUSSION THREAD

**Discussion by DmG-tech** [sourcelair/xterm.js#4319] [Q&A]
Xterm version: 4.19.0

Scroll go to top, when i typing and when i get output to console from remote server.
Terminal config:
```
cursorStyle: "underline",
cursorBlink: true,
theme: xtermjsTheme,
cols: 80,
```
I try: disableStdin:  on|off, rows: count, scrollback: count, macOptionClickForcesSelection: on|off.

I try use terminal:
```
    function initTerminal() {
        const protocol = "ws://";
        const host = document.location.host;
        const endpoint = contextPath + "/server/" + getServerId();
        const url = protocol + host + endpoint;

        websocket = new WebSocket(url);

        terminal = new Terminal(terminalOption);
        terminal.open(terminalContainer);
        terminal.focus();

        websocket.onmessage = function (data) {
            terminal.clear();
            terminal.write(data.data);
            terminal.scrollLines(99999999);
        };

        terminal.onData(function (data) {
            websocket.send(data);
            terminal.scrollLines(99999999);
        })

        terminal.onBinary(function (data) {
            websocket.send(data);
            terminal.scrollLines(99999999);
        })
    }
```
I try use - terminal.scrollToBottom() and terminal.scrollLines(99999999) - not working.

I try use:
```
var block = document.getElementsByClassName('xterm-viewport');
block.scrollTop = block.scrollHeight;
```
or
```
var block = document.getElementsByClassName('terminal xterm');
block.scrollTop = block.scrollHeight;
```
- not working.

I try use terminal with attach addons:
```
    function initTerminal() {
        const protocol = "ws://";
        const host = document.location.host;
        const endpoint = contextPath + "/server/" + getServerId();
        const url = protocol + host + endpoint;

        websocket = new WebSocket(url);
        websocket.onmessage = function (data) {
            terminal.clear();
        };

        terminal = new Terminal(terminalOption);
        const attachAddon = new AttachAddon.AttachAddon(websocket);
        terminal.loadAddon(attachAddon);
        terminal.open(terminalContainer);
        terminal.focus();
    }
```
- not working

This is a floating error, because sometimes autoscroll to buttom is work, but often is not work.

![изображение_2022-09-26_172603697](https://user-images.githubusercontent.com/53418595/192265017-be643913-3e0c-48be-8d29-4f1220040e08.png)

**Comment by Tyriar** [comment]
Are there any errors in the devtools console?

**Comment by DmG-tech** [comment]
No errors, only this message:
![изображение_2022-09-26_184436985](https://user-images.githubusercontent.com/53418595/192279640-cc031e39-44a2-43b8-9943-6cf4ca1c1efd.png)

**Comment by Tyriar** [comment]
I haven't seen anything like this, can you try reproduce it in the demo? https://github.com/xtermjs/xterm.js/wiki/Contributing#running-the-demo

**Comment by jerch** [comment]
@DmG-tech

Several things from my side:
- If correctly initialized, the terminal will do scroll-down automatically. Those `terminal.scrollLines` should not be needed.
- If your server sends byte data, `terminal.write(data.data);` will not work. You'd need to create an byte array first, e.g. `terminal.write(new Unint8Array(data.data));`. The amount of NULs in the console.log suggest, that you try something weird here. (Note that only UTF-8 is supported for byte data.)
- Have you resolved #4127 yet? Because if there is a fundamental misconception in your server setup, things will not work in the frontend either.

**Comment by illiav-j** [comment]
I'm having a similar problem. When I do a reset and then write multiple lines at once, there's a 50/50 chance it will be scrolled to the bottom.
I was able to reproduce it in the demo: illiav-j/xterm.js#1 (tested in Firefox and Chromium)

**Reply by jerch** [reply]
The `term.reset()` is a bit dangerous - it is a synchronous call, which disposes and re-creates half of the terminal's render chain. By the time you issue `term.write` the re-initialization might not be fully ready, so you run into false scroll offset calculation. `term.reset` is furthermore tricky, as it is not band synchronous to incoming parser data - thus should be used with great caution.

What helps here - either give `term.reset` some time to settle (e.g. `term.reset(); setTimeout(() => term.write(next_data), 50);`, or use a band synchronous VT function like RIS (e.g. `term.write('\x1bc'); term.write(next_data);`).

**Reply by jerch** [reply]
@Tyriar Not sure yet, if that behavior of `Terminal.reset()` is a recent regression (I remember having issues with it since v3.x). It seems after calling `reset` the terminal needs to go through at least one render cycle to get the offset calc right again (can mitigate the issue reliable by waiting ~20ms before writing new data). This prolly has to do with the viewport/scroll logic being bound to `requestAnimationFrame`.

Edit: Created #4603.

**Reply by illiav-j** [reply]
Thanks a lot @jerch! `term.write('\x1bc');` has solved the problem for me.

**Reply by illiav-j** [reply]
It seems `Terminal.clear()` also exhibits similar behaviour, which explains the problem OP had.
